---
title: "BIOSTAT-M280: Homework #1"
author: "Darrick Shen (UID:604946049) "
subtitle: Due Feb 2 @ 11:59PM
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(eval = TRUE)
```

## Q1. Git/GitHub

**No handwritten homework reports are accepted for this course.** We work with Git and GitHub. Efficient and abundant use of Git, e.g., frequent and well-documented commits, is an important criterion for grading your homework.

1. Apply for the [Student Developer Pack](https://education.github.com/pack) at GitHub using your UCLA email.

    This was done on 01/17/2018.

2. Create a **private** repository `biostat-m280-2018-winter` and add `Hua-Zhou` and `juhkim111` as your collaborators with write permission.

    This was done on 01/17/2018.

3. Top directories of the repository should be `hw1`, `hw2`, ... Maintain two branches `master` and `develop`. The `develop` branch will be your main playground, the place where you develop solution (code) to homework problems and write up report. The `master` branch will be your presentation area. Submit your homework files (R markdown file `Rmd`, `html` file converted from R markdown, all code and data sets to reproduce results) in `master` branch.

    This was done on 01/17/2018.

4. After each homework due date, teaching assistant and instructor will check out your master branch for grading. Tag each of your homework submissions with tag names `hw1`, `hw2`, ... Tagging time will be used as your submission time. That means if you tag your `hw1` submission after deadline, penalty points will be deducted for late submission.

    This was done on 01/17/2018. (Tagging will be done upon submission)
    
## Q2. Linux Shell Commands

The `35.227.165.60:/home/m280-data/hw1` folder contains a typical genetic data set in plink format. If interested, you can read plink documentation at <http://zzz.bwh.harvard.edu/plink/>. But it's definitely not necessary for this homework.

- `merge-geno.bim` contains information of each genetic marker (SNP). Each line is a SNP and has 6 fields:  
`Chromosome`, `SNP ID`, `Genetic Distance (morgan)`, `Base Pair Position (bp)`, `Allele 1`, `Allele 2`.
    ```{bash, echo=TRUE, eval=TRUE}
    head /home/m280-data/hw1/merge-geno.bim
    ```

- `merge-geno.fam` contains individual information. Each line is one individual and has 6 fields:  
`Family ID`, `Person ID`, `Father ID`, `Mother ID`, `Sex` coded as 1 (male) or 2 (female), `Affection Status`  
`Father ID = 0` means that person's father is not in this data set. Similarly `Mother ID` = 0 means that person's mother is not in this data set.
    ```{bash}
    head -20 /home/m280-data/hw1/merge-geno.fam
    ```

- `merge-geno.bed` contains genotypes of each individual in binary format. We don't need this file for this homework.

Please, do **not** put these data files into Git; they are huge. You even don't need to copy them into your directory. Just read from the data folder `/home/m280-data/hw1` directly.

Use Linux shell commands to answer following questions. 

1. How many persons are in the data set (statisticians call this `n`)?

    ```{bash}
    wc -l < /home/m280-data/hw1/merge-geno.fam
    # Here we are using wc -l to count number of rows in .fam file;
    ```
    There are 959 people in the merge-geno.fam data set.
    
    How many SNPs are in the data set (statisticians call this `p`)?

    ```{bash}
    wc -l < /home/m280-data/hw1/merge-geno.bim
    # Here we are using wc -l to count number of rows in .bim file;
    ```
    There are 8,348,674 SNPS in the merge-geno.bim data set.
    
2. Which chromosomes does this data set contain? 
    
    ```{bash}
    awk -F" " '!_[$1]++' /home/m280-data/hw1/merge-geno.bim | awk '{print $1}'
    # Here we are using the awk -F command as a filter for non-redundant  \
    # chromosomes in column 1;
    ```
    The chromosomes contained in this data set are: 3, 5, 7, 9, 11, 13, 15, 17, 19, 21
    
    How many SNPs are in each chromosome?

    Below, the total number of SNPS is listed, followed by the corresponding chromosome number.
    ```{bash}
    awk '{print $1}' /home/m280-data/hw1/merge-geno.bim | uniq -c 
    # Here we are using the awk command to print lines in column 1, but  \
    # results are piped into the uniq -c command to count the SNPS occurrences \
    # in each chromosome
    ```

3. MAP4 (microtubule-associated protein 4) is a gene on chromosome 3 spanning positions 47,892,180 bp -- 48,130,769 bp. How many SNPs are located within MAP4 gene?

    ```{bash}
    awk '{if ($1 == 3 && $4 >= 47892180 && $4 <= 48130769) print}'  \
    /home/m280-data/hw1/merge-geno.bim | wc -l
    # Here we use the awk command to filter for the chromsome specification  \
    # and bp range;
    ```

    There are a total of 894 SNPs located with the MAP4 gene.

4. Statistical geneticists often have to reformat a data set to feed into various analysis programs. For example, to use the Mendel software <http://www.genetics.ucla.edu/software/mendel>, we have to reformat the data set to be read by Mendel.

      - Mendel's SNP definition file is similar to the plink `bim` file but has format  
      `SNP ID`, `Chromosome`, `Base Pair Position`  
      with each field separated by a comma. Write a Linux shell command to convert `merge-geno.bim` to Mendel SNP definition file. The first few lines of the Mendel SNP definition file should look like
    
    ```{bash, echo=FALSE, eval=TRUE}
    awk 'BEGIN{print "    2.40 = FILE FORMAT VERSION NUMBER.  \
    \n8348674 = NUMBER OF SNPS LISTED HERE." }{print $2 "," $1 "," $4}'  \
    /home/m280-data/hw1/merge-geno.bim > mendel_snpdef.txt

    head -10 mendel_snpdef.txt
    ```
    The original merge-geno.bim data set has been converted into a Mendel's SNP definition file and stored
    in the text file named mendel_snpdef.txt. Above we preview 10 lines of the mendel_snpdef.txt file.
    
    - Mendel's pedigree file is similar to the plink `fam` file but has format  
    `Family ID`, `Person ID`, `Father ID`, `Mother ID`, `Sex` coded as M or F, `Twin Status`  
    with each field separated by a comma. Write a Linux shell command to convert `merge-geno.fam` to   Mendel pedigree file. Since twin status is not available in plink format, we put nothing for that field. Also Mendel limits Person ID to have length less than or equal to 8 characters, so we have to strip the string `T2DG` from the IDs. First few lines of the Mendel pedigree should look like
    
    ```{bash, echo=FALSE, eval=TRUE}
    awk '{OFS=","}{print $1,$2,$3,$4,$5,$6}'  \
    /home/m280-data/hw1/merge-geno.fam|sed 's/,1/,M/'| sed 's/,2/,F/'|  \
    sed 's/,0/,/g' |sed 's/T2DG//g' > mendel_ped.txt 
    
    head -20 mendel_ped.txt
    ```
    
    The original merge-geno.fam file has been converted into a plink 'fam' file and has been stored in
    a text file named mendel_ped.txt. Above we preview 20 lines fom the mendel_ped.txt file.

## Q3. R Batch Run

In class we discussed using R to organize simulation studies. 

1. Expand the [`runSim.R`](http://hua-zhou.github.io/teaching/biostatm280-2018winter/slides/02-linux/runSim.R) script to include arguments `seed` (random seed), `n` (sample size), `dist` (distribution) and `rep` (number of simulation replicates). When `dist="gaussian"`, generate data from standard normal; when `dist="t1"`, generate data from t-distribution with degree of freedom 1 (same as Cauchy distribution); when `dist="t5"`, generate data from t-distribution with degree of freedom 5. Calling `runSim.R` will (1) set random seed according to argument `seed`, (2) generate data according to argument `dist`, (3) compute the primed-indexed average estimator in class and the classical sample average estimator for each simulation replicate, (4) report the average mean squared error (MSE)
$$
  \frac{\sum_{r=1}^{\text{rep}} (\widehat \mu_r - \mu_{\text{true}})^2}{\text{rep}}
$$
for both methods.

    The runSim.R script was adjusted to include the arguments for seed, sample size, distribution type, and replicates. The script will be run in part 2) of this question.

2. Modify the [`autoSim.R`](http://hua-zhou.github.io/teaching/biostatm280-2018winter/slides/02-linux/autoSim.R) script to run simulations with combinations of sample sizes `nVals = seq(100, 500, by=100)` and distributions `distTypes = c("gaussian", "t1", "t5")` and write output to appropriately named files. Use `rep = 50`, and `seed = 280`. 
  
    The autoSim.R file will be called to run with arguments rep = 50 and seed = 280, evaluating each of distributions:            Gaussian, t1, t5, each at sample sizes of 100, 200, 300, 400, 500. The resultant averages for classical and Prime means
    will be stored seperately into an appropriately named text file (distribution_n_size_.txt).
    
    ```{bash, echo=FALSE, eval=TRUE} 
       Rscript autoSim.R
    ```

3. Write an R script to collect simulation results from output files and print average MSEs in a table of format

| $n$ | Method   | $t_1$ | $t_5$ | Gaussian |
|-----|----------|-------|-------|----------|
| 100 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 200 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 300 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 400 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |
| 500 | PrimeAvg |       |       |          |
|     | SampAvg  |       |       |          |

```{r, results = 'asis'}
dat <- lapply(Sys.glob("*_.txt"), read.table)

unl_dat <- matrix(unlist(dat), byrow = TRUE, ncol = 3)[, -1]; 
unl_dat2 <- unl_dat[c(6:15, 1:5), c(2, 1)]; 
unl_dat2_trans <- c(t(unl_dat2)) #transpose
matdat <- matrix(unl_dat2_trans, nrow = 10)

datfrmdat <- as.data.frame(matdat)
avgNames <- rep(c("PrimeAvg", "ClassAvg"), times = 5)
nObs <- rep(c(100, 200, 300, 400, 500), each = 2)
labeled_datfrmdat <- cbind(nObs, avgNames, datfrmdat)
colnames(labeled_datfrmdat) <- c("n", "Method", "t1", "t5", "Gaussian")

library(knitr)
knitr::kable(labeled_datfrmdat,  
             col.names = c("$n$             ",  
                           "            Method", "$t_1$", "$t_5$","Gaussian"))

```

